### PR TITLE
folly: enable folly MSVC (static)

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **folly/all**

Enable follly (static) msvc builds
(or detect what problems the ci will throw at us)

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/folly/all/conanfile.py [hunk 1]
```diff
@@ -58,26 +58,13 @@ def configure(self):
...
-        elif Version(self.version) >= "2020.08.10.00" and self.settings.compiler == "clang" and \
-                self.options.shared:
-            raise ConanInvalidConfiguration("Folly could not be built by clang as a shared library")
-
-        self._strip_options_requirements()
-
-    def _strip_options_requirements(self):
-        self.options["boost"].header_only = False
-        for boost_comp in self._required_boost_components:
-            setattr(self.options["boost"], "without_{}".format(boost_comp), False)
```
Comment (SpaceIm) [inline recipes/folly/all/conanfile.py:80]: I believe this part shouldn't be removed (only fix the typo in function name ^^), a recipe should be able to work out of the box, so require specific features of a dependency (which could have been a specific recipe if boost was modularized).

**Comment by prince-chrismc** [comment]
To "fix" the build you will need to make a complicated patch to fix the bad code in folly. https://github.com/conan-io/conan-center-index/pull/2535#discussion_r503553880

File: recipes/folly/all/CMakeLists.txt [hunk 2]
```diff
@@ -4,4 +4,8 @@ project(cmake_wrapper)
 include(conanbuildinfo.cmake)
 conan_basic_setup()
 
-add_subdirectory("source_subfolder")
+if(MSVC)
+    add_compile_options(/std:c++latest)
```
Comment (prince-chrismc) [inline recipes/folly/all/CMakeLists.txt:8]: Maybe?

```suggestion
    add_compile_options(/std:c++latest /permissive-)
```

**Comment by madebr** [comment]
This built completely fine on MSVC2017 and MSVC2019 on my Windows box.
It's weird that it fails on c3i.
This make me thinking that the MSVC toolchain on c3i is not up to date.

**Comment by prince-chrismc** [comment]
> This built completely fine on MSVC2017 and MSVC2019 on my Windows box.
> It's weird that it fails on c3i.
> This make me thinking that the MSVC toolchain on c3i is not up to date.

It failed for me locally.... I do not update agressively

**Comment by madebr** [comment]
> > This built completely fine on MSVC2017 and MSVC2019 on my Windows box.
> > It's weird that it fails on c3i.
> > This make me thinking that the MSVC toolchain on c3i is not up to date.
> 
> It failed for me locally.... I do not update agressively

See https://github.com/serge-sans-paille/pythran/issues/990#issuecomment-493126611
When you goddbolt the code there, the example fails with the same error.
The error disappears when you add `/std:c++latest`.

For folly, I think the problem is with the constexpr variable `make_atomic_ref`.
See https://github.com/facebook/folly/blob/67ed8482ff3e0e4a99d74d89f0224bc67bc632e3/folly/synchronization/AtomicRef.h#L135-L147

Maybe we should try replacing the functor with a plain function. e.g.:
```c++
template <typename T,
    std::enable_if_t<
        is_trivially_copyable_v<T> && sizeof(T) == sizeof(std::atomic<T>) &&
            alignof(T) == alignof(std::atomic<T>),
        int> = 0>>
FOLLY_INLINE_VARIABLE atomic_ref<T> make_atomic_ref(T& ref) {
    return atomic_ref<T>{ref};
}
```
*(untested)*

**Comment by prince-chrismc** [comment]
closes #3171

**Comment by prince-chrismc** [comment]
This was fixed with #3711 

 https://conan.io/center/folly?os=Windows&tab=configuration

**Comment by prince-chrismc** [comment]
@madebr can you close this PR please, it's fixed already

**Comment by madebr** [comment]
@prince-chrismc 
Thanks for checking.

**Pull request closed without merging**